### PR TITLE
Change servant-js and servant-foreign to use text

### PR DESCRIPTION
**Pull request by arianvp**
Changes servant-js and servant-foreign to use Text instead of String

### CODE REVIEW

**Comment by arianvp** [comment]
I'd like to know how others feel about this change.  The pragmatic me feels that String should be avoided at all cost and that Text is a lot cleaner. But wondering what other people think of it.

File: servant-js/test/Servant/JSSpec.hs [hunk 1]
```diff
@@ -81,12 +85,16 @@ spec = describe "Servant.JQuery" $ do
     axiosSpec
     --angularSpec    AngularCustom
 
+
```
Comment (jkarni) [inline servant-js/test/Servant/JSSpec.hs]: Maybe signatures here?

File: servant-js/src/Servant/JS/Internal.hs [hunk 2]
```diff
@@ -87,7 +88,34 @@ defCommonGeneratorOptions = CommonGeneratorOptions
 -- Couldn't work out how to handle zero-width characters.
 --
 -- @TODO: specify better default function name, or throw error?
-toValidFunctionName :: String -> String
+toValidFunctionName :: Text -> Text
+-- @TODO: Cons text
```
Comment (jkarni) [inline servant-js/src/Servant/JS/Internal.hs]: What's this TODO about?

File: servant-js/src/Servant/JS/Internal.hs [hunk 3]
```diff
@@ -87,7 +88,34 @@ defCommonGeneratorOptions = CommonGeneratorOptions
...
+                    , Set.titlecaseLetter
+                    , Set.modifierLetter
+                    , Set.otherLetter
+                    , Set.letterNumber ]
+    remainderOK   = firstLetterOK
+               <> [ Set.nonSpacingMark
+                  , Set.spacingCombiningMark
+                  , Set.decimalNumber
+                  , Set.connectorPunctuation ]
+{-
```
Comment (jkarni) [inline servant-js/src/Servant/JS/Internal.hs]: We can remove the comment, no?

File: servant-js/test/Servant/JSSpec.hs (hunk 1 above)
Comment (arianvp) [inline servant-js/test/Servant/JSSpec.hs]: Yes will add them.   They're not exported though as they're just internal helper functions.  But I guess it's prettier to add a signature to them

File: servant-js/src/Servant/JS/Internal.hs (hunk 2 above)
Comment (arianvp) [inline servant-js/src/Servant/JS/Internal.hs]: The `Cons text` TODO should be removed. it concerns a piece of code that doesn't exist anymore.

The other `TODO` as already there in the original version so I left it untouched.

File: servant-js/src/Servant/JS.hs [hunk 4]
```diff
@@ -142,6 +145,8 @@ writeJSForAPI :: (HasForeign api, GenerateList (Foreign api))
               -> JavaScriptGenerator -- ^ js code generator to use (angular, vanilla js, jquery, others)
               -> FilePath -- ^ path to the file you want to write the resulting javascript code into
               -> IO ()
+
+-- TODO Data.Text
```
Comment (jkarni) [inline servant-js/src/Servant/JS.hs]: And this one?

File: servant-examples/digest-auth/digest-auth.hs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{-#LANGUAGE DataKinds, TypeOperators, OverloadedStrings #-}
+module Main where
```
Comment (arianvp) [inline servant-examples/digest-auth/digest-auth.hs]: I accidentally added this file in one of the commits.   I'll remove it again as it's not suppoted to be in this pull.

File: servant-js/src/Servant/JS/Internal.hs [hunk 6]
```diff
@@ -87,7 +88,34 @@ defCommonGeneratorOptions = CommonGeneratorOptions
...
+      setFirstChar x `T.cons` T.filter remainder xs 
+    Nothing -> "_"
+  where
+    setFirstChar c = if firstChar c then c else '_'
+    firstChar c = prefixOK c || any (Set.member c) firstLetterOK
+    remainder c = prefixOK c || any (Set.member c) remainderOK
+    -- Valid prefixes
+    prefixOK c = c `elem` ['$','_']
+    -- Unicode character sets
+    firstLetterOK = [ Set.lowercaseLetter
```
Comment (jkarni) [inline servant-js/src/Servant/JS/Internal.hs]: (I know this wasn't introduced in your change but) I find it a little more intuitive to do a union of sets here and below (and probably more efficient), and then get rid of the `any`s above

**Comment by jkarni** [comment]
LGTM

**Comment by dredozubov** [comment]
+1 on this

**Comment by jkarni** [comment]
@arianvp could you add a line to the changelog?

**Pull request merged**